packages/cozy-doctypes/src/Folder.js

```diff
@@ -0,0 +1,105 @@
+const Application = require('./Application')
+const CozyFile = require('./File')
+
+/**
+ * Class representing the folder model.
+ * @extends CozyFile
+ */
+class CozyFolder extends CozyFile {
+  /**
+   * Create a folder with a reference to the given document
+   * @param  {String}  path     Folder path
+   * @param  {Object}  document
+   * @return {Object}  Folder document
+   */
+  static async createFolderWithReference(path, document) {
+    const collection = this.cozyClient.collection(CozyFile.doctype)
+    const dirId = await collection.ensureDirectoryExists(path)
+    await collection.addReferencesTo(document, [
+      {
+        _id: dirId
+      }
+    ])
+
+    const { data: dirInfos } = await collection.get(dirId)
+
+    return dirInfos
+  }
+
+  /**
+   * Returns a "Magic Folder", given its id
+   * @param  {String} id Magic Folder id. Could be:
+   * * `CozyFolder.refs.ADMINISTRATIVE`
+   * * `CozyFolder.refs.PHOTOS`
+   * * `CozyFolder.refs.PHOTOS_BACKUP`
+   * * `CozyFolder.refs.PHOTOS_UPLOAD`
+   * @param {String} path Default path to use if magic folder does not exist
+   * @return {Object} Folder document
+   */
+  static async ensureMagicFolder(id, path) {
+    const magicFolderDocument = {
+      _type: Application.doctype,
+      _id: id
+    }
+    const folders = await CozyFolder.getReferencedFolders(magicFolderDocument)
+    const existingMagicFolder = folders.length ? folders[0] : null
+
+    if (existingMagicFolder) return existingMagicFolder
+
+    const magicFoldersValues = Object.values(CozyFolder.refs)
+    if (!magicFoldersValues.includes(id)) {
+      throw new Error(
+        `Cannot create Magic folder with id ${id}. Allowed values are ${magicFoldersValues.join(
+          ', '
+        )}.`
+      )
+    }
+
+    if (!path) {
+      throw new Error('Magic folder default path must be defined')
+    }
+
+    return CozyFolder.createFolderWithReference(path, magicFolderDocument)
+  }
+
+  /**
+   * Returns an array of folder referenced by the given document
+   * @param  {Object}  document  Document to get references from
+   * @return {Array}             Array of folders referenced with the given
+   * document
+   */
+  static async getReferencedFolders(document) {
+    const { included } = await this.cozyClient
+      .collection(CozyFile.doctype)
+      .findReferencedBy(document)
+    return included.filter(
+      folder => !/^\/\.cozy_trash/.test(folder.attributes.path)
+    )
+  }
+
+  /**
+   * Returns an unique folder referenced with the given reference. Creates it
+   * if it does not exist.
+   * @param  {String}  path      Path used to create folder if the referenced
+   * folder does not exist.
+   * @param  {String}  reference
+   * @return {Objet}             Folder referenced with the give reference
+   */
+  static async getOrCreateFolderWithReference(path, document) {
```

Comment:
Shouldn't we rename this to `ensureFolderWithReference` to echo the method `ensureMagicFolder` ?